File: frontend/src/lib/api/media.ts
import { api } from "./client";
import type { Media, PresignedUpload } from "@/types";

export function presignUpload(contentType: string) {
  return api.post<PresignedUpload>("/api/v1/media/presign", { content_type: contentType });
}

export interface CaptureProof {
  sha256?: string;
  captured_at?: string;
  captured_lat?: number;
  captured_lng?: number;
}

export function registerMedia(
  input: {
    object_key: string;
    content_type: string;
    related_type?: "incident" | "result";
    related_id?: string;
  } & CaptureProof,
) {
  return api.post<Media>("/api/v1/media/register", input);
}

/** Resolves a batch of media IDs (e.g. an incident's media_ids) to their
 * viewable records/URLs in one call. */
export function getMediaBatch(ids: string[]) {
  if (ids.length === 0) return Promise.resolve<Media[]>([]);
  return api.get<Media[]>(`/api/v1/media?ids=${ids.map(encodeURIComponent).join(",")}`);
}

/** Presigns, PUTs the file straight to object storage, then registers the
 * resulting object as media. The API server never sees the file bytes.
 * `proof` carries the watermark/hash metadata for a tamper-evident
 * capture (see lib/media/watermark.ts) -- omitted for ordinary photo/
 * video/voice attachments that don't need it. */
export async function uploadFile(
  file: File,
  related?: { related_type: "incident" | "result"; related_id?: string },
  proof?: CaptureProof,
): Promise<Media> {
  const presigned = await presignUpload(file.type);
  // Longer bound than the JSON API calls -- a real photo/video upload on a
  // slow-but-working connection can legitimately take a while, so this
  // only exists to stop a genuinely dead connection attempt from hanging
  // indefinitely (see client.ts's REQUEST_TIMEOUT_MS for why that matters
  // to the offline queue specifically), not to cut off a slow-but-real upload.
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 60000);
  let res: Response;
  try {
    res = await fetch(presigned.upload_url, {
      method: "PUT",
      headers: { "Content-Type": file.type },
      body: file,
      signal: controller.signal,
    });
  } finally {
    clearTimeout(timeout);
  }
  if (!res.ok) {
    // An HTTP-level rejection from storage (expired presigned URL, wrong
    // content-type, a transient 5xx) -- deliberately a plain Error, not
    // ApiError, so callers' isRetriable check (lib/offline/queue.ts)
    // treats this the same as a dropped connection and queues the file
    // for retry instead of silently registering a Media record for bytes
    // that were never actually stored.
    throw new Error(`Upload to storage failed (${res.status})`);
  }
  return registerMedia({
    object_key: presigned.object_key,
    content_type: file.type,
    ...related,
    ...proof,
  });
}

File: backend/internal/domain/media.go
package domain

import (
	"context"
	"time"
)

type RelatedType string

const (
	RelatedIncident RelatedType = "incident"
	RelatedResult   RelatedType = "result"
)

type Media struct {
	ID          string      `bson:"_id,omitempty" json:"id"`
	ObjectKey   string      `bson:"object_key" json:"object_key"`
	URL         string      `bson:"url" json:"url"`
	ContentType string      `bson:"content_type" json:"content_type"`
	UploadedBy  string      `bson:"uploaded_by" json:"uploaded_by"`
	RelatedType RelatedType `bson:"related_type,omitempty" json:"related_type,omitempty"`
	RelatedID   string      `bson:"related_id,omitempty" json:"related_id,omitempty"`
	CreatedAt   time.Time   `bson:"created_at" json:"created_at"`
}

type MediaRepository interface {
	Create(ctx context.Context, m *Media) error
	FindByID(ctx context.Context, id string) (*Media, error)
	FindByIDs(ctx context.Context, ids []string) ([]*Media, error)
}

// PresignedUpload is issued to a client so it can upload a file directly to
// object storage without proxying bytes through the API.
type PresignedUpload struct {
	UploadURL string `json:"upload_url"`
	ObjectKey string `json:"object_key"`
	PublicURL string `json:"public_url"`
	ExpiresIn int    `json:"expires_in_seconds"`
}

// ObjectStore is implemented by the storage adapter (Cloudflare R2 / S3).
type ObjectStore interface {
	PresignPutURL(ctx context.Context, objectKey, contentType string) (*PresignedUpload, error)
	PublicURL(objectKey string) string
}
